text/0000-io-fs-2.1.md

```diff
@@ -0,0 +1,541 @@
+```
+
+## Enhancing soft link support
+
+Currently the `std::fs` module provides a `soft_link` and `read_link` function,
+but there is no method of doing other soft link related tasks such as:
+
+* Testing whether a file is a soft link
+* Reading the metadata of a soft link, not what it points to
+
+The following APIs will be added to `std::fs`:
+
+```rust
+/// Returns the metadata of the file pointed to by `p`, and this function,
+/// unlike `metadata` will **not** follow soft links.
+pub fn soft_link_metadata<P: AsRef<Path>>(p: P) -> io::Result<Metadata>;
+
+impl Metadata {
+    /// Tests whether this metadata is for a soft link or not.
+    pub fn is_soft_link(&self) -> bool;
+}
+```
+
+## Binding `realpath`
+
+There's a [long-standing issue][realpath] that the unix function `realpath` is
+not bound, and this RFC proposes adding the following API to the `fs` module:
+
+[realpath]: https://github.com/rust-lang/rust/issues/11857
+
+```rust
+/// Canonicalizes the given file name to an absolute path with all `..`, `.`,
+/// and soft link components resolved.
+///
+/// On unix this function corresponds to the return value of the `realpath`
+/// function, and on Windows this corresponds to the `GetFullPathName` function.
+///
+/// Note that relative paths given to this function will use the current working
+/// directory as a base, and the current working directory is not managed in a
+/// thread-local fashion, so this function may need to be synchronized with
+/// other calls to `env::change_dir`.
+pub fn canonicalize<P: AsRef<Path>>(p: P) -> io::Result<PathBuf>;
+```
+
+## Tweaking `PathExt`
+
+Currently the `PathExt` trait is unstable, yet it is quite convenient! The main
+motivation for its `#[unstable]` tag is that it is unclear how much
+functionality should be on `PathExt` versus the `std::fs` module itself.
+Currently a small subset of functionality is offered, but it is unclear what the
+guiding principle for the contents of this trait are.
+
+This RFC proposes a few guiding principles for this trait:
+
+* Only read-only operations in `std::fs` will be exposed on `PathExt`. All
+  operations which require modifications to the filesystem will require calling
+  methods through `std::fs` itself.
+
+* Some inspection methods on `Metadata` will be exposed on `PathExt`, but only
+  those where it logically makes sense for `Path` to be the `self` receiver. For
+  example `PathExt::len` will not exist (size of the file), but
+  `PathExt::is_dir` will exist.
+
+Concretely, the `PathExt` trait will be expanded to:
+
+```rust
+pub trait PathExt {
+    fn exists(&self) -> bool;
+    fn is_dir(&self) -> bool;
+    fn is_file(&self) -> bool;
+    fn is_soft_link(&self) -> bool;
+    fn metadata(&self) -> io::Result<Metadata>;
+    fn soft_link_metadata(&self) -> io::Result<Metadata>;
+    fn canonicalize(&self) -> io::Result<PathBuf>;
+    fn read_link(&self) -> io::Result<PathBuf>;
```

Comment:
Currently on Windows `read_link` recursively resolves all intermediate symbolic links. Is the behavior going to be changed so that it only reads a single layer?